File: backend-python/app/routes/click.py
from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy import select
from sqlalchemy.orm import Session
from ..auth import get_current_user
from ..database import get_db
from ..models import Card, Transaction, User
from ..rate_limit import check_click_limit
from ..schemas import ClickBody

router = APIRouter(prefix='/api/click', tags=['click'])

@router.post('')
def click(body: ClickBody, request: Request, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    check_click_limit(request, user.id)
    try:
        import uuid
        card_id = uuid.UUID(body.card_id)
    except ValueError:
        raise HTTPException(400, 'Valid card_id is required')
    with db.begin():
        card = db.scalar(select(Card).where(Card.id == card_id, Card.user_id == user.id).with_for_update())
        if not card:
            raise HTTPException(404, 'Card not found')
        card.balance += 1
        db.add(Transaction(receiver_card_id=card.id, amount=1, type='click'))
        db.flush()
        new_balance = card.balance
    return {'success': True, 'card_id': str(card_id), 'new_balance': str(new_balance)}

File: backend-node/src/routes/click.js
import express from 'express';
import { pool } from '../db.js';
import { clickRateLimit } from '../rate_limit.js';

const router = express.Router();

router.post('/', clickRateLimit, async (req, res, next) => {
  const cardId = String(req.body?.card_id || '');
  if (!/^[0-9a-fA-F-]{36}$/.test(cardId)) return res.status(400).json({ error: 'Valid card_id is required' });
  const client = await pool.connect();
  try {
    await client.query('BEGIN');
    const result = await client.query(
      `UPDATE cards SET balance = balance + 1.00
       WHERE id = $1 AND user_id = $2
       RETURNING id, balance`,
      [cardId, req.user.id]
    );
    if (!result.rows[0]) {
      await client.query('ROLLBACK');
      return res.status(404).json({ error: 'Card not found' });
    }
    await client.query(
      `INSERT INTO transactions (sender_card_id, receiver_card_id, amount, type)
       VALUES (NULL, $1, 1.00, 'click')`,
      [cardId]
    );
    await client.query('COMMIT');
    return res.json({ success: true, card_id: cardId, new_balance: result.rows[0].balance });
  } catch (error) {
    await client.query('ROLLBACK');
    return next(error);
  } finally {
    client.release();
  }
});

export default router;
